### PR TITLE
sequence_task_1 AMarmusevich

### PR DESCRIPTION
**Pull request by marmusevich**

### CODE REVIEW

File: sequence_task_1/student/include/impl.h [hunk 1]
```diff
@@ -19,32 +20,121 @@ template<class Container>
...
+    /** Member traits */
+    /*
+     // simple ...
+    typedef Song         value_type;
+    typedef Song&        reference;
+    typedef const Song&  const_reference;
+    typedef size_t	     size_type;
+    typedef ptrdiff_t    difference_type;
+    */
+    typedef typename Container::value_type       value_type;
```
Comment (smay1613) [inline sequence_task_1/student/include/impl.h:32]: Please prefer 'using'
https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rt-using

File: sequence_task_1/student/include/impl.h [hunk 2]
```diff
@@ -19,32 +20,121 @@ template<class Container>
...
+    {
+
+    }
+
+    /** Assignment from any reversible sequence container */
+    //any container? Cheking needed AnyContainer is supported in compile time, like static_assert. Out of scope.
+    template<class AnyContainer>
+    StaticPlaylist& operator=(const AnyContainer& ac)
+    {
+        m_tracklist.clear();
```
Comment (smay1613) [inline sequence_task_1/student/include/impl.h:78]: You don't need to do it manually. Prefer other methods.

File: sequence_task_1/student/include/impl.h [hunk 3]
```diff
@@ -19,32 +20,121 @@ template<class Container>
...
+        m_tracklist.insert(m_tracklist.begin(), ac.begin(), ac.end());
+        return *this;
+    }
+
+    StaticPlaylist& operator=( const StaticPlaylist& other )
```
Comment (smay1613) [inline sequence_task_1/student/include/impl.h:83]: You can express this method via template above

File: sequence_task_1/student/include/impl.h [hunk 4]
```diff
@@ -19,32 +20,121 @@ template<class Container>
...
+        }
+        return *this;
+    }
+
+    /** Add track from initializer */
     template<class... Args>
-    const Song& play(Args&&... songData);
+    const Song& play(Args&&... songData)
+    {
+        m_tracklist.emplace_back(songData...);
```
Comment (smay1613) [inline sequence_task_1/student/include/impl.h:96]: Use std::forward
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f19-for-forward-parameters-pass-by-tp-and-only-stdforward-the-parameter

File: sequence_task_1/student/include/impl.h [hunk 5]
```diff
@@ -19,32 +20,121 @@ template<class Container>
...
+
+    }
+
+    /** Assignment from any reversible sequence container */
+    //any container? Cheking needed AnyContainer is supported in compile time, like static_assert. Out of scope.
+    template<class AnyContainer>
+    StaticPlaylist& operator=(const AnyContainer& ac)
+    {
+        m_tracklist.clear();
+        m_tracklist.insert(m_tracklist.begin(), ac.begin(), ac.end());
```
Comment (smay1613) [inline sequence_task_1/student/include/impl.h:79]: **Not effective for vector**

File: sequence_task_1/student/include/impl.h [hunk 6]
```diff
@@ -19,32 +20,121 @@ template<class Container>
...
 
-    /** @todo Skip to the next track in playlist, remove current */
-    void switchNext();
+    /** Skip to the next track in playlist, remove current */
+    void switchNext()
+    {
+        if(hasTracks())
+        {
+            //first is current, so erase first, second will be current
+            m_tracklist.erase(m_tracklist.begin());
```
Comment (smay1613) [inline sequence_task_1/student/include/impl.h:119]: **Not effective for vector**

**Review by smay1613 — CHANGES_REQUESTED** [review]

File: sequence_task_1/student/include/impl.h (hunk 1 above)
Comment (marmusevich) [reply sequence_task_1/student/include/impl.h]: accept

File: sequence_task_1/student/include/impl.h (hunk 2 above)
Comment (marmusevich) [reply sequence_task_1/student/include/impl.h]: accept

File: sequence_task_1/student/include/impl.h (hunk 3 above)
Comment (marmusevich) [reply sequence_task_1/student/include/impl.h]: accept, remove

File: sequence_task_1/student/include/impl.h (hunk 4 above)
Comment (marmusevich) [reply sequence_task_1/student/include/impl.h]: accept

File: sequence_task_1/student/include/impl.h (hunk 6 above)
Comment (marmusevich) [reply sequence_task_1/student/include/impl.h]: reworked

File: sequence_task_1/student/include/impl.h (hunk 5 above)
Comment (marmusevich) [reply sequence_task_1/student/include/impl.h]: reworked

File: sequence_task_1/student/include/impl.h [hunk 7]
```diff
@@ -59,13 +61,13 @@ class StaticPlaylist
     //any container? Cheking needed AnyContainer is supported in compile time, like static_assert. Out of scope.
     template<class AnyContainer>
     StaticPlaylist(const AnyContainer& ac)
-    : m_tracklist(ac.begin(), ac.end())
+    : m_tracklist(ac.rbegin(), ac.rend())
```
Comment (smay1613) [inline sequence_task_1/student/include/impl.h:64]: Yeah, right! :)

File: sequence_task_1/student/include/impl.h [hunk 8]
```diff
@@ -107,16 +99,15 @@ class StaticPlaylist
...
     }
 
     /** Skip to the next track in playlist, remove current */
     void switchNext()
     {
         if(hasTracks())
         {
-            //first is current, so erase first, second will be current
-            m_tracklist.erase(m_tracklist.begin());
+            (void)m_tracklist.pop_back();
```
Comment (smay1613) [inline sequence_task_1/student/include/impl.h:110]: why did you casted it to void?

Comment (marmusevich) [reply sequence_task_1/student/include/impl.h:110]: shown that i ignore return value, code stile in my project

Comment (smay1613) [reply sequence_task_1/student/include/impl.h:110]: Looks like a bug, pop_back returns nothing

**Pull request closed without merging**